contracts/clients/src/Erc20Compressor.ts

```diff
@@ -0,0 +1,390 @@
+
+    const [
+      blsPublicKeyRegistryContract,
+      addressRegistryContract,
+      aggregatorUtilitiesContract,
+    ] = await Promise.all([
+      blsPublicKeyRegistryFactory.deploy(),
+      addressRegistryFactory.deploy(),
+      aggregatorUtilitiesFactory.deploy(),
+    ]);
+
+    const erc20ExpanderFactory = new ERC20ExpanderFactory(signer);
+
+    const erc20ExpanderContract = await erc20ExpanderFactory.deploy(
+      blsPublicKeyRegistryContract.address,
+      addressRegistryContract.address,
+      aggregatorUtilitiesContract.address,
+    );
+
+    return new Erc20Compressor(
+      erc20ExpanderContract,
+      new BlsPublicKeyRegistryWrapper(blsPublicKeyRegistryContract),
+      new AddressRegistryWrapper(addressRegistryContract),
+      aggregatorUtilitiesContract,
+    );
+  }
+
+  static async connectOrDeploy(
+    signerOrFactory: Signer | SafeSingletonFactory,
+    salt: ethers.utils.BytesLike = ethers.utils.solidityPack(["uint256"], [0]),
+  ): Promise<Erc20Compressor> {
+    const factory = await SafeSingletonFactory.from(signerOrFactory);
+
+    const [blsPublicKeyRegistry, addressRegistry, aggregatorUtilities] =
+      await Promise.all([
+        BlsPublicKeyRegistryWrapper.connectOrDeploy(factory, salt),
+        AddressRegistryWrapper.connectOrDeploy(factory, salt),
+        factory.connectOrDeploy(AggregatorUtilitiesFactory, [], salt),
+      ]);
+
+    const erc20ExpanderContract = await factory.connectOrDeploy(
+      ERC20ExpanderFactory,
+      [
+        blsPublicKeyRegistry.registry.address,
+        addressRegistry.registry.address,
+        aggregatorUtilities.address,
+      ],
+      salt,
+    );
+
+    return new Erc20Compressor(
+      erc20ExpanderContract,
+      blsPublicKeyRegistry,
+      addressRegistry,
+      aggregatorUtilities,
+    );
+  }
+
+  static async connectIfDeployed(
+    signerOrProvider: SignerOrProvider,
+    salt: ethers.utils.BytesLike = ethers.utils.solidityPack(["uint256"], [0]),
+  ): Promise<Erc20Compressor | undefined> {
+    const factoryViewer = await SafeSingletonFactoryViewer.from(
+      signerOrProvider,
+    );
+
+    const [
+      blsPublicKeyRegistryAddress,
+      addressRegistryAddress,
+      aggregatorUtilitiesAddress,
+    ] = await Promise.all([
+      factoryViewer.calculateAddress(BLSPublicKeyRegistryFactory, [], salt),
+      factoryViewer.calculateAddress(AddressRegistryFactory, [], salt),
+      factoryViewer.calculateAddress(AggregatorUtilitiesFactory, [], salt),
+    ]);
+
+    const erc20Expander = await factoryViewer.connectIfDeployed(
+      ERC20ExpanderFactory,
+      [
+        blsPublicKeyRegistryAddress,
+        addressRegistryAddress,
+        aggregatorUtilitiesAddress,
+      ],
+      salt,
+    );
+
+    if (!erc20Expander) {
+      return undefined;
+    }
+
+    return await Erc20Compressor.wrap(erc20Expander);
+  }
+
+  async compress(blsPublicKey: PublicKey, operation: Operation) {
```

Comment:
My opinion is definitely not.

In general, I think it is much better to mostly only use type definitions where they are needed. If the compiler infers the type correctly, then you should generally leave it alone.

This makes writing and modifying code much easier as the code is more fluid. Types often change but the usage remains consistent, and the compiler just checks that it flows through correctly. In the alternative where extra type annotations are added, you need to update those annotations too, which is extra friction.

For the readability side, you can get the type from hovering - I don't think there is a significant benefit to the reader. It can even be better for the reader since that type only shows up if the reader wants to see it, otherwise they don't need to be distracted by it.

Here's some more information on this point of view: https://www.youtube.com/watch?v=RmGHnYUqQ4k.

Definitely aware there are other perspectives here and happy to talk about it further.